File: 166_StackUsingQueue.cpp
/*
Problem: Implement Stack using Queues
Description: Implement stack operations (push, pop, top) using queue(s).

Approach: Use one queue. For push, add element and rotate queue to make new element front.
*/

#include <bits/stdc++.h>
using namespace std;

class MyStack {
private:
    queue<int> q;
    
public:
    MyStack() {}
    
    void push(int x) {
        int size = q.size();
        q.push(x);
        for (int i = 0; i < size; i++) {
            q.push(q.front());
            q.pop();
        }
    }
    
    int pop() {
        int top = q.front();
        q.pop();
        return top;
    }
    
    int top() {
        return q.front();
    }
    
    bool empty() {
        return q.empty();
    }
};

int main() {
    MyStack stack;
    stack.push(1);
    stack.push(2);
    cout << "Top: " << stack.top() << endl;    // Output: 2
    cout << "Pop: " << stack.pop() << endl;    // Output: 2
    cout << "Empty: " << stack.empty() << endl; // Output: 0
    
    return 0;
}

/*
Time Complexity: push O(n), pop/top/empty O(1)
Space Complexity: O(n)
*/

File: 166_StackUsingQueue.py
"""
Problem: Implement Stack using Queues
Description: Implement stack operations (push, pop, top) using queue(s).

Approach: Use one queue. For push, add element and rotate queue to make new element front.
"""

from collections import deque

class MyStack:
    def __init__(self):
        self.q = deque()
    
    def push(self, x: int) -> None:
        size = len(self.q)
        self.q.append(x)
        for _ in range(size):
            self.q.append(self.q.popleft())
    
    def pop(self) -> int:
        return self.q.popleft()
    
    def top(self) -> int:
        return self.q[0]
    
    def empty(self) -> bool:
        return len(self.q) == 0

def main():
    stack = MyStack()
    stack.push(1)
    stack.push(2)
    print(f"Top: {stack.top()}")      # Output: 2
    print(f"Pop: {stack.pop()}")      # Output: 2
    print(f"Empty: {stack.empty()}")  # Output: False

if __name__ == "__main__":
    main()

"""
Time Complexity: push O(n), pop/top/empty O(1)
Space Complexity: O(n)
"""
